modules/apps/users-admin/users-admin-impl/src/main/java/com/liferay/users/admin/internal/model/listener/UserModelListener.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.users.admin.internal.model.listener;
+
+import com.liferay.petra.sql.dsl.DSLQueryFactoryUtil;
+import com.liferay.portal.kernel.exception.ModelListenerException;
+import com.liferay.portal.kernel.exception.RequiredUserException;
+import com.liferay.portal.kernel.model.BaseModelListener;
+import com.liferay.portal.kernel.model.ModelListener;
+import com.liferay.portal.kernel.model.User;
+import com.liferay.portal.kernel.model.UserConstants;
+import com.liferay.portal.kernel.model.UserTable;
+import com.liferay.portal.kernel.service.UserLocalService;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Pei-Jung Lan
+ */
+@Component(service = ModelListener.class)
+public class UserModelListener extends BaseModelListener<User> {
+
+	@Override
+	public void onBeforeRemove(User user) throws ModelListenerException {
+		if (user.getType() != UserConstants.TYPE_SERVICE_ACCOUNT) {
+			return;
+		}
+
+		int count = _userLocalService.dslQueryCount(
+			DSLQueryFactoryUtil.countDistinct(
+				UserTable.INSTANCE.userId
+			).from(
+				UserTable.INSTANCE
+			).where(
+				UserTable.INSTANCE.type.eq(UserConstants.TYPE_SERVICE_ACCOUNT)
+			));
+
+		if (count == 1) {
+			throw new ModelListenerException(new RequiredUserException());
+		}
```

Comment:
Could we change this so that only the default service account cannot be removed? I know the default userName `service_account_user` is loosely defined, but perhaps we could harden the definition of the default service account to have a more trictly specified name and then prevent deletion (or change) of that specific name? The name is important since it should be safe to use as an external identity in portable configurations so it is really API.

Thoughts?

Reply:
Hi @rotty3000, the idea for preventing the deletion of the lasts service account is to ensure that there will always be a service account available. I might be missing something, but is there anything special about the default service account that we want to prevent it from being removed? I'm asking because I'm wondering if it's necessary to have the additional logic to distinguish the default service account from the other service accounts.